flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java

```diff
@@ -0,0 +1,257 @@
+          value.group().rewrittenFiles());
+    } else {
+      LOG.info(
+          LogUtil.MESSAGE_PREFIX + "Rewriting {} files",
+          tableName,
+          taskName,
+          taskIndex,
+          ctx.timestamp(),
+          value.group().rewrittenFiles().size());
+    }
+
+    try (TaskWriter<RowData> writer = writerFor(value)) {
+      try (DataIterator<RowData> iterator = readerFor(value)) {
+        while (iterator.hasNext()) {
+          writer.write(iterator.next());
+        }
+
+        Set<DataFile> dataFiles = Sets.newHashSet(writer.dataFiles());
+        value.group().setOutputFiles(dataFiles);
+        out.collect(
+            new ExecutedGroup(
+                value.table().currentSnapshot().snapshotId(),
+                value.groupsPerCommit(),
+                value.group()));
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(
+              LogUtil.MESSAGE_PREFIX + "Rewritten files {} from {} to {}",
+              tableName,
+              taskName,
+              taskIndex,
+              ctx.timestamp(),
+              value.group().info(),
+              value.group().rewrittenFiles(),
+              value.group().addedFiles());
+        } else {
+          LOG.info(
+              LogUtil.MESSAGE_PREFIX + "Rewritten {} files to {} files",
+              tableName,
+              taskName,
+              taskIndex,
+              ctx.timestamp(),
+              value.group().rewrittenFiles().size(),
+              value.group().addedFiles().size());
+        }
+      } catch (Exception ex) {
+        LOG.info(
+            LogUtil.MESSAGE_PREFIX + "Exception rewriting datafile group {}",
+            tableName,
+            taskName,
+            taskIndex,
+            ctx.timestamp(),
+            value.group(),
+            ex);
+        ctx.output(TaskResultAggregator.ERROR_STREAM, ex);
+        errorCounter.inc();
+        abort(writer, ctx.timestamp());
+      }
+    } catch (Exception ex) {
+      LOG.info(
+          LogUtil.MESSAGE_PREFIX + "Exception creating compaction writer for group {}",
+          tableName,
+          taskName,
+          taskIndex,
+          ctx.timestamp(),
+          value.group(),
+          ex);
+      ctx.output(TaskResultAggregator.ERROR_STREAM, ex);
+      errorCounter.inc();
+    }
+  }
+
+  private TaskWriter<RowData> writerFor(PlannedGroup value) {
+    String formatString =
+        PropertyUtil.propertyAsString(
+            value.table().properties(),
+            TableProperties.DEFAULT_FILE_FORMAT,
+            TableProperties.DEFAULT_FILE_FORMAT_DEFAULT);
+    RowDataTaskWriterFactory factory =
+        new RowDataTaskWriterFactory(
+            value.table(),
+            FlinkSchemaUtil.convert(value.table().schema()),
+            value.splitSize(),
+            FileFormat.fromString(formatString),
+            value.table().properties(),
+            null,
+            false);
+    factory.initialize(subTaskId, attemptId);
```

Comment:
This is the question of how flexible we want the compaction flow to be.
Creating the factory once during the open method would mean, that the compaction will always write with the old schema - even if the table is updated in the meantime

Reply:
MonitorSource provides the possibility to have a standalone job to run the maintenance for a given table.
While changing the table schema/spec behind a streaming ingest job will not cause issues, if we start compacting new files with the old schema we will start losing data from the new colums, as the will not be written to the compacted files. This should be prevented by either adding guards against it, or by making sure we are able to handle dynamic changes.
Discussion with users around the DynamicSink convinced me, that this is a valid concern, and we should move forward with dynamic solutions whenever it is possible.

@mxm: What are your thoughts?

Reply:
We could always make this configurable, but I would suggest to default to the dynamic variant. The reason mainly being that it doesn't hurt to use the up-to-date table metadata with the current Iceberg sink. Plus, it enables to run in standalone mode, where we definitely want to refresh the table metadata. Note that being static with Flink is much worse than with Spark, because of the long-running nature of Flink streaming applications. Once we attach the compaction to the Dynamic Sink, not being dynamic is out of question.